packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts

```diff
@@ -131,21 +149,33 @@ export class Pipeline extends cdk.Construct implements cpapi.IPipeline {
       service: 'codepipeline',
       resource: this.pipelineName
     });
+
+    for (const stage of props.stages || []) {
+      this.addStage(stage);
+    }
   }
 
   /**
-   * Convenience method for creating a new {@link Stage},
-   * and adding it to this Pipeline.
+   * Adds a Stage to this Pipeline.
    *
-   * @param name the name of the newly created Stage
-   * @param props the optional construction properties of the new Stage
-   * @returns the newly created Stage
+   * @param stage the newly created Stage to add to this Pipeline
+   * @param props optional Stage insertion properties
    */
-  public addStage(name: string, props?: CommonStageProps): Stage {
-    return new Stage(this, name, {
-      pipeline: this,
-      ...props,
-    });
+  public addStage(stage: cpapi.IStage, props: StageInsertionProps = {}): Pipeline {
```

Comment:
Fluent APIs are a bit tricky and we try to avoid them because if a user extends `Pipeline` the API will not work as expected:

```ts
const p = new MyPipeline(...);
p.addStage(...) // returns Pipeline and not MyPipeline
```

And jsii doesn't support generics to be able to express this properly.

Reply:
Removed the fluent API. It was not possible to have it with the new way of specifying Stages (without creating them explicitly first) anyway.